# Import Library

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

## Content

### 1. Import Data
### 2. If-Statment using FUNCTION for 'price_range'
### 3. If-Statment using loc() FUNCTION for 'price_range_loc'
### 4. If-Statement using For-Loops for 'busiest_day'
### 5. If-Statement using For-Loops for 'busiest_days'
### 6. If-Statement using For-Loops for 'busiest_period_of_day'
### 7. Export Data

#### 1. Import 'orders_products_merged'

In [2]:
# String for Main Project Folder
path = r'C:\Users\ifari\12-2022 Instacart Basket Analysis'

In [3]:
# String to import Data Folder containing orders_products_merged.pkl
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))

In [4]:
# Drop merge column
df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['_merge'])

In [5]:
# First 5 rows of 'orders_products_merged' dataframe
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0   2539329        1             1                   2              8   
1   2398795        1             2                   3              7   
2    473747        1             3                   3             12   
3   2254736        1             4                   4              7   
4    431534        1             5                   4             15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices  
0         Soda        77              7     9.0  
1         Soda        77              7     9.0  
2         Soda        77              7     9.0  
3         Soda        77              7     9.0  
4         Soda        77              7     9.0

In [6]:
# Selecting the first 1 million rows
df = df_ords_prods_merged[:1000000]

#### 2. If-Statment using FUNCTION to create 'price_range' column in subset

In [7]:
# Defining the price_label function
def price_label(row):
    
  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [8]:
# Create new column 'price_range'
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\ifari\AppData\Local\Temp\ipykernel_12672\1105986058.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [9]:
# Frequency of the values in the 'price_range' column
df['price_range'].value_counts(dropna = False)

Mid-range product    756450
Low-range product    243550
Name: price_range, dtype: int64

In [10]:
# Maxmimum price
df['prices'].max()

14.8

> There are no 'High range product'. This is confirmed because the highest price of a product is 14.8 which is a 'Mid range product', concludingthat there aren’t any high-range products within the subset.

#### 3. If-Statment using loc() FUNCTION to create 'price_range_loc' column in the 'orders_products_merged' dataframe

In [11]:
# Defining the price_label_loc function - High
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [12]:
# Defining the price_label_loc function - Mid
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [13]:
# Defining the price_label_loc function - Low
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [14]:
# Frequency of the values in the 'price_range_loc' column
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: price_range_loc, dtype: int64

#### 4. If-Statement using For-Loops to create 'busiest_day' column in the 'orders_products_merged' dataframe

In [15]:
# Frequency of the values in the 'orders_day_of_week' column
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [16]:
# Loop checking and assigning values
result_1 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0:
    result_1.append("Busiest day")
  elif value == 4:
    result_1.append("Least busy")
  else:
    result_1.append("Regularly busy")

In [17]:
# Define the 'busiest_day' function
df_ords_prods_merged['busiest_day'] = result_1

In [18]:
# Frequency of the values in the 'busiest_day' column
df_ords_prods_merged['busiest_day'].value_counts(dropna = False)

Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: busiest_day, dtype: int64

In [19]:
# Length of the 'busiest_day' column
len(result_1)

32404859

In [20]:
# Dimensions of the 'order_produces_merged' dataframe
df_ords_prods_merged.shape

(32404859, 15)

> The frequency of the new column 'busiest_day' matches the frequency of the 'orders_day_of_week'.

#### 5. If-Statement using For-Loops to create 'busiest_days' column in the 'orders_products_merged' dataframe

In [21]:
# Frequency of the values in the 'orders_day_of_week' column
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: orders_day_of_week, dtype: int64

In [22]:
# Loop checking and assigning values
result_2 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 3 or value == 4:
    result_2.append("Slowest days")
  else:
    result_2.append("Regular days")

In [23]:
# Define the 'busiest_days' function
df_ords_prods_merged['busiest_days'] = result_2 

In [24]:
# Frequency of the values in the 'busiest_days' column
df_ords_prods_merged['busiest_days'].value_counts(dropna = False)

Regular days    12916111
Busiest days    11864412
Slowest days     7624336
Name: busiest_days, dtype: int64

In [25]:
# Length of the 'busiest_day' column
len(result_2)

32404859

In [26]:
# Dimensions of the 'order_produces_merged' dataframe
df_ords_prods_merged.shape

(32404859, 16)

In [27]:
# First 5 rows of 'orders_products_merged' dataframe
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0   2539329        1             1                   2              8   
1   2398795        1             2                   3              7   
2    473747        1             3                   3             12   
3   2254736        1             4                   4              7   
4    431534        1             5                   4             15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day  busiest_days  
0  Regularly busy  Regular days  
1  Regularly busy  Slowest days  
2  Regularly busy  Slowest days  
3      Least busy  Slowest days  
4      Least busy  Slowest days

#### 6. If-Statement using For-Loops to create 'busiest_period_of_day' column in the 'orders_products_merged' dataframe

In [28]:
# Frequency of the values in the 'time_of_order' column
df_ords_prods_merged['time_of_order'].value_counts(dropna = False)

10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: time_of_order, dtype: int64

In [29]:
# Loop checking and assigning values
result_3 = []

for value in df_ords_prods_merged["time_of_order"]:
  if value >= 0 and value <= 6: 
     result_3.append("Fewest Orders")
  elif value == 23:
    result_3.append("Fewest Orders")
  elif value >= 9 and value <= 16:
    result_3.append("Most Orders")
  else:
    result_3.append("Average Orders")

In [30]:
# Define the 'busiest_period_of_day' function
df_ords_prods_merged['busiest_period_of_day'] = result_3 

In [31]:
# Frequency of the values in the 'busiest_period_of_day' column
df_ords_prods_merged['busiest_period_of_day'].value_counts(dropna = False)

Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: busiest_period_of_day, dtype: int64

In [32]:
# Length of the 'busiest_period_of_day' column
len(result_3)

32404859

In [33]:
# Dimensions of the 'order_produces_merged' dataframe
df_ords_prods_merged.shape

(32404859, 17)

In [34]:
# First 5 rows of 'orders_products_merged' dataframe
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  time_of_order  \
0   2539329        1             1                   2              8   
1   2398795        1             2                   3              7   
2    473747        1             3                   3             12   
3   2254736        1             4                   4              7   
4    431534        1             5                   4             15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                    15.0         196                  1          1   
2                    21.0         196                  1          1   
3                    29.0         196                  1          1   
4                    28.0         196                  1          1   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda        77              7     9.0  Mid-range product   
1         Soda        77              7     9.0  Mid-range product   
2         Soda        77              7     9.0  Mid-range product   
3         Soda        77              7     9.0  Mid-range product   
4         Soda        77              7     9.0  Mid-range product   

      busiest_day  busiest_days busiest_period_of_day  
0  Regularly busy  Regular days        Average Orders  
1  Regularly busy  Slowest days        Average Orders  
2  Regularly busy  Slowest days           Most Orders  
3      Least busy  Slowest days        Average Orders  
4      Least busy  Slowest days           Most Orders

#### 7. Export Dataframe

In [35]:
# Export data in pickle format
df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_derived.pkl'))